### ISSUE TITLE
Document and version lockfile fields semantics

### ISSUE THREAD

**Issue by staktrace** [openfare/openfare#5]
Currently it looks like the `OPENFARE.lock` fields are basically defined by the rust structs in the codebase and (de-)serialized using serde. While this makes initial implementation fast, it would be good consider the long-term usage of this file. Specifically, different repositories will have `OPENFARE.lock` files generated by different versions of the tool that will be long-lived (i.e. may not be updated for years/decades). Developers may wish to use other third-party tools to generate and process the files. As such, it would be good to:

1. Ensure the file format is versioned, so as to allow forwards- and backwards-compatibility. That is, old versions of the `openfare` binary should gracefully detect and fail when attempting to read `OPENFARE.lock` files which were generated by future versions of `openfare` binaries. And they should gracefully handle older versions of `OPENFARE.lock` files that were generated by older versions, ideally without bumping up the version number.
2. Document the allowed fields in the file and their meanings.

Additionally as discussed in the matrix chat, it would be good use a file format that allowed for in-file comments as well ([hjson](https://hjson.github.io/) was mentioned, although after looking at I'm not such a big fan of it).

Basically this issue amounts to "standardize the OPENFARE.lock file format" to the point where there can exist independent tools that can operate on it.

**Comment by willhansen** [comment]
" 2. Document the allowed fields in the file and their meanings. "

I haven't used it before, but [JSON Schema](https://json-schema.org/) looks like it's made for this sort of thing ([examples](https://json-schema.org/learn/miscellaneous-examples.html)).

There's a [rust crate](https://crates.io/crates/jsonschema) for it too.  
From what I could see, the licenses are BSD and MIT, so it should be good from a licensing angle.

We could probably throw on an `openfare lock validate [<file>]` command too.

For comments, we could probably just go with `"_comment": <string> ` entries allowed anywhere in the lock file that are stripped out immediately after parsing.

I'd be willing to take a shot at implementing it.

**Comment by rndhouse** [comment]
That sounds great to me!

The [schemars](https://crates.io/crates/schemars) crate might be useful. It could potentially help avoid repeating the same structure fields in two different locations.

The `lock validate` sub-command is probably a good idea.

> For comments, we could probably just go with "_comment": <string> entries allowed anywhere in the lock file that are stripped out immediately after parsing.

Not sure about comments. I'm concerned about comments which contradict field values. I suppose as long as it's clear that they will be ignored by openfare.

---

> 1. Ensure the file format is versioned

They are now versioned. (But the version number won't mean much before openfare v1.0.0)

**Comment by willhansen** [comment]
@rndhouse 

I agree about comments.  They really just leave room for added confusion in a configuration file that is supposed to be legally binding.  I'll leave them off for now.

Just noticed you've got payment distributions defined per-plan in the example lock files.  Do you envision a use-case for different distributions based on plans?

![Selection_028](https://user-images.githubusercontent.com/2918280/162555919-4d1e519d-c514-4283-b0e8-46ec6a59350b.png)

  I would have guessed a "payment_distribution" object on the same level as "plans" and "payees" would be simpler.  (or even folding number of shares into each payee object, to avoid excessively custom payee labels at a slight cost to readability)

**Comment by rndhouse** [comment]
@willhansen I have thought recently about refactoring the shares scheme to the top level of a lock file. And then referencing a particular scheme from within a plan.

I think a change will be necessary because it's conceivable to have many different payment plans. And it wouldn't be a good idea to  unnecessarily repeat the shares scheme each time.

**Comment by willhansen** [comment]
Factoring out the distributions makes sense, as does having multiple payment plans, but I'm not following the need for multiple distributions between contributors.  It's like saying "Money is split equally between contributors, unless the user goes with the month-to-month plan, then Jim doesn't get a cut", or I'm missing an obvious use case.

**Comment by rndhouse** [comment]
You're right. I think supporting multiple shares scheme is over engineering. Will go with one for now. Thanks for the push back.

**Comment by rndhouse** [comment]
@willhansen Commit 7d076293d15565f13f0c0963cd5e531b4e2434f8 moves `shares` to the top level of the lock json.

Commit f5ece45cecde761b415219431c35ef772a0ba773 replaces the `payments.total` fields would become `price`.

**Comment by rndhouse** [comment]
I've merged https://github.com/openfare/openfare/pull/9 by @willhansen . There's probably more to add, but it's a fantastic start. I'm going to close this issue and we can always open more for further work on this topic.
